src/v/config/constraints.cc

```diff
@@ -45,6 +52,500 @@ std::ostream& operator<<(std::ostream& os, const constraint_t& constraint) {
+    // the cluster one. An undefined topic value or cluster value could not
+    // match the other, this implies that the constraint is not satisfied.
+    if (!topic_val || !cluster_val) {
+        return false;
+    }
+
+    if (topic_val != cluster_val) {
+        vlog(
+          constraints_log.error,
+          "Constraints failure[does not match the cluster property {}]: "
+          "topic property {}.{}, value {}",
+          cluster_property,
+          topic(),
+          topic_property,
+          topic_val);
+        return false;
+    }
+
+    return true;
+}
+
+/**
+ * Returns true if the topic-level value satisfies the constraint.
+ * \param topic_val: the topic value
+ * \param constraint: the constraint to evaluate
+ * \param property_name: name of the cluster-level property
+ * \param cluster_opt: the value from the cluster-level property
+ */
+template<typename T>
+bool validate_value(
+  const std::optional<T>& topic_val,
+  const constraint_t& constraint,
+  const model::topic& topic,
+  const std::string_view& topic_property,
+  const std::string_view& cluster_property,
+  const std::optional<T> cluster_opt = std::nullopt) {
+    if (constraint.name == cluster_property) {
+        return ss::visit(
+          constraint.flags,
+          [&topic_val,
+           &topic,
+           &topic_property,
+           &cluster_property,
+           &cluster_opt](const constraint_enabled_t enabled) {
+              return matches_cluster_property_value(
+                topic_val,
+                cluster_opt,
+                enabled,
+                topic,
+                topic_property,
+                cluster_property);
+          },
+          [&topic_val, &topic, &topic_property](const auto range) {
+              return within_range(topic_val, range, topic, topic_property);
+          });
+    }
+    return false;
+}
+
+template<typename T>
+bool validate_value(
+  const tristate<T>& tri,
+  const constraint_t& constraint,
+  const model::topic& topic,
+  const std::string_view& topic_property,
+  const std::string_view& cluster_property,
+  const std::optional<T> cluster_opt = std::nullopt) {
+    if (tri.is_disabled()) {
+        return false;
+    }
+
+    return validate_value(
+      tri.get_optional(),
+      constraint,
+      topic,
+      topic_property,
+      cluster_property,
+      cluster_opt);
+}
+} // namespace
+
+bool topic_config_satisfies_constraint(
+  const cluster::topic_configuration& topic_cfg,
+  const constraint_t& constraint) {
+    auto partition_count_opt = std::make_optional(topic_cfg.partition_count);
+    auto replication_factor_opt = std::make_optional(
+      topic_cfg.replication_factor);
+    bool ret
+      = validate_value(
+          partition_count_opt,
+          constraint,
+          topic_cfg.tp_ns.tp,
+          kafka::topic_property_partition_count,
+          config::shard_local_cfg().default_topic_partitions.name())
```

Comment:
Since you know the name of the constraint, can you do a lookup to get the related cluster config rather than having to check every one?  Because only one cluster config is going to match the provided constraint, correct?

Reply:
Or is this large conditional predicated on trying to extract the correct topic property?

Reply:
Could you write out a switch/if statement that checks the name in `constraint`?

Reply:
So, earlier iterations had the switch/if statement that you are referring to but there were a couple "clean-code" problems with it, such as: (1) It's a giant switch/if. One for each property name (2) `topic_config_satisfies_constraint` and `constraint_clamp_topic_config` would both need the giant switch/if. De-duplicating the code led to some cascading changes for calling validate_value vs clamp_value and optional vs tristate which also looked un-clean. So I opted for the current approach. I am happy to re-visit this approach if it is a larger concern.

Reply:
A shorter answer: the constraint name is checked within `validate_value`, so the correct configs are checked against their corresponding constraint and it looks nicer than a big if statement.